openlibrary/core/booknotes.py

```diff
@@ -0,0 +1,143 @@
+        query += ' group by work_id order by cnt desc limit $limit'
+        return list(oldb.query(query, vars={
+            'limit': limit,
+            'since': since
+        }))
+
+
+    @classmethod
+    def count_total_booksnotes_by_user(cls, username):
+        """Counts the (int) total number of books logged by this `username`
+        """
+        oldb = db.get_db()
+        data = {'username': username}
+        query = "SELECT count(*) from booknotes WHERE username=$username"
+        return oldb.query(query, vars=data)['count']
+
+
+    @classmethod
+    def get_patron_booknote(cls, username, work_id):
+        note = cls.get_patron_booknotes(username, work_id=work_id)
+        return note and note[0]
+
+    @classmethod
+    def get_patron_booknotes(cls, username, work_id=None, search=None, limit=100, page=1):
+        """
+        By default, get all a patron's booknotes. if work_id, get only that booknote
+
+        return:
+        """
+        oldb = db.get_db()
+        page = int(page) if page else 1
+        data = {
+            'username': username,
+            'work_id': work_id,
+            'limit': limit,
+            'offset': limit * (page - 1),
+            'search': search
+        }
+        query = "SELECT * from booknotes WHERE username=$username "
+        if work_id:
+            query += "AND work_id=$work_id "
+        if search:
+            query += "AND notes LIKE '%$search%' "
+        query += "LIMIT $limit OFFSET $offset"
+        return list(oldb.query(query, vars=data))
+
+
+    @classmethod
+    def add(cls, username, work_id, notes, edition_id=None):
+        """Insert or update booknote. Create a new booknote if one doesn't
+        exist, or gracefully update the record otherwise.
+
+        return: the updates booknote record from the db.
+        """
+        oldb = db.get_db()
+        data = {
+            "work_id": work_id,
+            "username": username,
+            "notes": notes,
+            "edition_id": edition_id
+        }
+        records = cls.get_patron_booknotes(username, work_id=work_id)
+        if not records:
+            return oldb.insert(
+                'booknotes',
+                username=username,
+                work_id=work_id,
+                notes=notes,
+                edition_id=edition_id
+            )
+        return oldb.update(
+            'booknotes',
+            where="work_id=$work_id AND username=$username",
+            notes=notes,
+            edition_id=edition_id
+        )
+
+
+    @classmethod
+    def remove(cls, username, work_id):
+        """Remove a patron's specific booknote by work_id.
+
+        Technical note: work_id is not an optional argument and
+        intentionally does not default to None (to reduce
+        accidents/risk), however if one passes None as a value to
+        work_id, this method will remove all booknotes for a patron
+        (useful for a patron who may decide to close their account.
```

Comment:
Is this accurate?  `work_id` is a primary key field, so it is not nullable by default.

Casting `None` to an int (at line 134) will cause a TypeError, as well.

If any of this should be changed, I can work on it while I make booknotes edition-level.